include/highfive/bits/H5Node_traits_misc.hpp

```diff
@@ -105,6 +106,24 @@ inline Group NodeTraits<Derivate>::createGroup(const std::string& group_name) {
     return group;
 }
 
+template <typename Derivate>
+inline Group NodeTraits<Derivate>::createGroup(const std::string& group_path,
+                                               bool recursive) {
+    if (recursive) {
+        auto split = std::find(group_path.begin(), group_path.end(), '/');
+
+        if (split != group_path.end()) {
+            std::string cur_node_name(group_path.begin(), split);
+            Group cur_g = exist(cur_node_name)? getGroup(cur_node_name)
+                                              : createGroup(cur_node_name);
+            return cur_g.createGroup(std::string(split + 1, group_path.end()),
+                                     true);
+        }
+    }
+    // No recursion
+    return _createGroup(group_path);
```

Comment:
I'd insist on not having the recursive parameter and write this:

    const auto split = group_path.find_first_of('/');
    if (split != std::string::npos) {
        const auto cur_node_name = group_path.substr(0, split);
        Group cur_g = exist(cur_node_name)? getGroup(cur_node_name)
            : createGroup(cur_node_name);
        return cur_g.createGroup(group_path.substr(split + 1));
    }
    else
        return _createGroup(group_path);

That's less code, fewer parameters and less documentation.

You also have to deal with the case of a group name with a trailing slash or two consecutive slashes. None of the cases should be accepted, it should throw and in my opinion no group should be created if the group name is ill formed (for exception safety reasons).

Reply:
I agree on the trailing slash(es) that should throw. However, based on HDF5's documentation as referenced above it should be concluded that any number of repeated slashes are legal and should be simply interpreted as a single slash.

Reply:
On second thought, I am even doubting about the trailing slash(es). The unix command `mkdir /path/to/new/dir/` would also be acceptable, I see no reason why it should be any different here.

Reply:
The trailing `/` still has to be handled correctly by the implementation, and right now it's not the case as it tries to create a leaf group with an empty name.
I wouldn't take other tools as an example as I'm aware that some UNIX tools behave differently when there's a trailing / and the end. For example `ln -s foo bar`and `ln -s foo bar/` don't mean the same. On the other hand, h5 command line tools accept trailing slashes, so I won't oppose it.

Reply:
Hi guys. Thanks for your input.
My reasons to prefer recursive off by default are:

- we dont change hdf5 lib way of working which could mislead people
- dont break compat
- mimick better filesystem mkdir

in terms of code size we only saved 1 `if`. I'd care more for the whole indentation, which might be possible to improve.

@hernando you raised an important point, that we should leave things clean if it fails. I also like the idea of validating the path before blindly start the process. I'll try to find some time to work on it

Reply:
I was about to reason in favor of having recursive on by default, but before that I wanted to test what happens if you try to create a group with a slash in the name in the current master. To my surprise, I've discovered that there is a test for this, that means that nested groups are probably handled by the HDF5 lib already and this change is unnecessary. Look in test_high_five_base.cpp 247

Reply:
And according to my last observation, maybe that property is on by default.

Reply:
in test_high_five_base.cpp 247 they have already GROUP_CREATED by line 245. 
I'm pretty sure recursive is off by default, that's also why so many people complain about this and think hdf5 doesnt do it (including me).

Reply:
I didn't pay attention to that. Then I'll write back what I didn't send yesterday.

I'd insist on making recursive the default and only behaviour (either with H5Pset_create_intermediate_group or manually implemented, I don't care). Consider that slashes are not allowed in group names, so this code:

    auto foo = file.createGroup("/root/foo");

fails at the moment. Adding the recursive parameter won't change the fact that any casual reading of code using HighFive which hasn't read the documentation (and very little people read docs these days, they only read stackoverflow) will be surprised about this behaviour, moreover if you consider that h5py creates the intermediate groups.

My point here is that:

> * we dont change hdf5 lib way of working which could mislead people

Is a non-reason. You are not going to mislead people, when in fact people are complaining that `file.createGroup("/root/foo")` doesn't work. Having to add the additional bool parameter doesn't make the code intent readable either, because there's no way to understand what the unnamed boolean parameter (which are always discouraged) without reading the code. I always advocate for self-documenting code, which one you think is more obvious `file.createGroup("/root/foo")` or  `file.createGroup("/root/foo", true)`?

> * dont break compat

It's a another non-reason, as you're relaxing the preconditions, not changing the semantics. Old code will still work the same. The only difference is that what before was an exception that probably nobody was handling, but rather avoiding, now would work by default.

Reply:
Couldn't agree more, the recursive version should be the default and only. If a user specifically wants to avoid this, she/he can always call `exist` before calling `createGroup`. The same argumentation holds for `exists` by the way.

Reply:
Guys, I think we are making a big deal out of this option. I don't care, one way or the other. But for some reason many libs don't do recursive magic by default to be safe. And if people don't read the docs where a signature has a `recursive=false`, I'm sorry maybe they shouldn't program :p
@adevress @tristan0x What is your view? Yes (recursive default) or No? (we dont need more arguments)

Reply:
My 2 cents: it depends where one puts HighFive on the scale between "wrapper around HDF5 C interface" and "also, simplify usage for most often usecases at the cost of diverging from HDF5 C interface defaults".
My personal (slight) preference would be the latter one.
Especially for `exists`, I can't imagine why would I want an exception because of a missing intermediate group (for `createGroup` and `mkdir` there are different approaches out there).